pkg/planner/core/planbuilder.go

```diff
@@ -1286,10 +1287,23 @@ func getPossibleAccessPaths(ctx base.PlanContext, tableHints *hint.PlanHints, in
 				publicPaths = append(publicPaths, path)
 				continue
 			}
+			hasTiCIIndex = hasTiCIIndex || index.IsFulltextIndexOnTiCI()
 			path := &util.AccessPath{Index: index}
 			publicPaths = append(publicPaths, path)
 		}
 	}
+	var ticiIndexPaths []*util.AccessPath
+	if hasTiCIIndex {
+		// FTS_MATCH_XXX can only be executed in TiCI engine.
+		// So we need to store it here and try to add it later if the USE_INDEX hint delete any of them.
+		// The removal of the unhinted TiCI index paths will be done after we dicide the availability of each index.
+		ticiIndexPaths = make([]*util.AccessPath, 0, len(publicPaths))
+		for _, path := range publicPaths {
+			if path.Index != nil && path.Index.IsFulltextIndexOnTiCI() {
```

Comment:
why don't we add it in previous loop, which i mean L1290, when we see index is TiCIIndex, is must be a public path, so just collecting it into ticiIndexPaths in a same loop is profitable.